chain/client/src/sync/state.rs

```diff
@@ -1237,35 +1300,57 @@ fn request_part_from_external_storage(
     part_id: u64,
     download: &mut DownloadStatus,
     shard_id: ShardId,
+    sync_hash: CryptoHash,
     epoch_id: &EpochId,
     epoch_height: EpochHeight,
     num_parts: u64,
     chain_id: &str,
-    requests_remaining: Arc<AtomicI32>,
+    semaphore: Arc<Semaphore>,
     external: ExternalConnection,
+    state_parts_arbiter_handle: &ArbiterHandle,
+    state_parts_mpsc_tx: Sender<StateSyncGetPartResult>,
 ) {
-    if !allow_request(&requests_remaining) {
+    if !download.run_me.swap(false, Ordering::SeqCst) {
+        tracing::info!(target: "sync", %shard_id, part_id, "run_me is already false");
         return;
-    } else {
-        if !download.run_me.swap(false, Ordering::SeqCst) {
-            tracing::info!(target: "sync", %shard_id, part_id, "run_me is already false");
-            return;
-        }
     }
     download.state_requests_count += 1;
     download.last_target = None;
 
     let location =
         external_storage_location(chain_id, epoch_id, epoch_height, shard_id, part_id, num_parts);
-    let download_response = download.response.clone();
-    near_performance_metrics::actix::spawn("StateSync", {
-        async move {
-            let result = external.get_part(shard_id, &location).await;
-            finished_request(&requests_remaining);
-            let mut lock = download_response.lock().unwrap();
-            *lock = Some(result.map_err(|err| err.to_string()));
+
+    match semaphore.try_acquire_owned() {
+        Ok(permit) => {
+            if state_parts_arbiter_handle.spawn({
```

Comment:
Currently `sync_shards_download_parts_status()` is driving the downloads in the sense that it sets `run_me` and records the timestamp of the beginning of the download and decides when and if a download needs to be retried.
Seems that driving the downloads and the downloads can be put together in some abstraction, hiding this complexity from `StateSync`.
What do you think about that?

Reply:
I see that the header sync retry logic is done in the same way. On the other hand, I would love to see the state sync being handled by the StateSync actor and `sync_shards_download_parts_status()` only check on a status flag updated by the StateSync actor.

After we do this refactoring, the State sync actor will handle the retry logic and also send notifications to the client actor to update the sync status.